File: AlgoExpert/Graphs/Medium/River Sizes.java
import java.util.*;

class Program {
  public static List<Integer> riverSizes(int[][] matrix) {
    // Write your code here.
		ArrayList<Integer> sizes = new ArrayList<>();
		boolean[][] visited = new boolean[matrix.length][matrix[0].length];
		
		for (int i = 0; i < matrix.length; i++) {
			for (int j = 0; j < matrix[i].length; j++) {
				if (visited[i][j]) continue;
				traverseNode(i, j, matrix, visited, sizes);
			}
		}		
		
    return sizes;
  }
	
	public static void traverseNode(int i, int j, int[][] matrix, boolean[][] visited, ArrayList<Integer> sizes) {
		int currentRiverSize = 0;
		Stack<Integer[]> nodesToExplore = new Stack<>();
		nodesToExplore.push(new Integer[] {i, j});
		
		while (!nodesToExplore.isEmpty()) {
			Integer[] node = nodesToExplore.pop();
			i = node[0];
			j = node[1];
			
			if (visited[i][j])
				continue;
			visited[i][j] = true;
			
			if (matrix[i][j] == 0)
				continue;
			
			currentRiverSize++;
			
			ArrayList<Integer[]> unvisitedNeighbors = getUnvisitedNeighbors(i, j, matrix, visited);
			for (Integer[] neighbor: unvisitedNeighbors) {
				nodesToExplore.push(neighbor);
			}
		}
		
		if (currentRiverSize > 0)
			sizes.add(currentRiverSize);
	}
	
	public static ArrayList<Integer[]> getUnvisitedNeighbors(int i, int j, int[][] matrix, boolean[][] visited) {
		ArrayList<Integer[]> unvisitedNeighbors = new ArrayList<>();
		
		if (i > 0 && !visited[i - 1][j])
			unvisitedNeighbors.add(new Integer[] {i - 1, j});
		
		if (i < matrix.length - 1 && !visited[i + 1][j])
			unvisitedNeighbors.add(new Integer[] {i + 1, j});
			
		if (j > 0 && !visited[i][j - 1])
			unvisitedNeighbors.add(new Integer[] {i, j - 1});
		
		if (j < matrix[i].length - 1 && !visited[i][j + 1])
			unvisitedNeighbors.add(new Integer[] {i, j + 1});
			
		return unvisitedNeighbors;
	}
}

File: Python/AlgoExpert/Graphs/Medium/riverSizes.py
def riverSizes(matrix):
    # Write your code here.
    visited = [[False for j in range(len(matrix[0]))] for i in range(len(matrix))]
    result = []

    for i in range(len(matrix)):
        for j in range(len(matrix[i])):
            if visited[i][j]:
                continue
                
            if matrix[i][j] == 1:
                result.append(find_river(matrix, visited, i, j))
            

    return result


def find_river(matrix, visited, row, col):
    to_visit = [(row, col)]
    visited[row][col] = True
    river_length = 0
    
    while to_visit:
        current = to_visit.pop(0)
        row = current[0]
        col = current[1]
        river_length += 1
        
        if row - 1 >= 0:
            if matrix[row - 1][col] == 1 and not visited[row - 1][col]:
                to_visit.append((row - 1, col))
                visited[row - 1][col] = True
        if col - 1 >= 0:
            if matrix[row][col - 1] == 1 and not visited[row][col - 1]:
                to_visit.append((row, col - 1))
                visited[row][col - 1] = True
        if row + 1 < len(matrix):
            if matrix[row + 1][col] == 1 and not visited[row + 1][col]:
                to_visit.append((row + 1, col))
                visited[row + 1][col] = True
        if col + 1 < len(matrix[row]):
            if matrix[row][col + 1] == 1 and not visited[row][col + 1]:
                to_visit.append((row, col + 1))
                visited[row][col + 1] = True
        
    return river_length



